File: 백준/Gold/10026. 적록색약/적록색약.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline
n = int(input())

d = [(0,1), (0, -1), (1,0), (-1,0)]
def dfs(x, y):
    visited[x][y] = True
    color = graph[x][y]
    for dx, dy in d:
        X, Y = x + dx, y + dy
        if (0 <= X < n) and (0 <= Y < n):
            if visited[X][Y] == False and graph[X][Y] == color:
                dfs(X, Y)

cnt, cnt2 = 0, 0
graph = []
for i in range(n):
    graph.append(list(map(str, input().strip())))
visited = [[False] * n for _ in range(n)]

for i in range(n):
    for j in range(n):
        if visited[i][j]==False:
            dfs(i, j)
            cnt += 1

visited = [[False] * n for _ in range(n)]
for i in range(n):
    for j in range(n):
        if graph[i][j] == 'G':
            graph[i][j] = 'R'

for i in range(n):
    for j in range(n):
        if visited[i][j] == False:
            dfs(i, j)
            cnt2 += 1
            
print(cnt, cnt2)

File: 백준/Gold/10026. 적록색약/적록색약.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
	static int n;
    static char[][] graph;
    static boolean[][] visited;
    static int[][] deltas = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        n = Integer.parseInt(br.readLine());
        graph = new char[n][n];
        visited = new boolean[n][n];
        
        for (int i = 0; i < n; i++) {
            String line = br.readLine().trim();
            for (int j = 0; j < n; j++) {
                graph[i][j] = line.charAt(j);
            }
        }
        
        int cnt = 0;
        int cnt2 = 0;
        
        // 정상적인 경우
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j]) {
                    dfs(i, j, graph[i][j]);
                    cnt++;
                }
            }
        }
        
        // 방문 배열 초기화
        visited = new boolean[n][n];
        
        // 적록색약인 경우 'G'를 'R'로 변환
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 'G') {
                    graph[i][j] = 'R';
                }
            }
        }
        
        // 적록색약인 경우 탐색
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visited[i][j]) {
                    dfs(i, j, graph[i][j]);
                    cnt2++;
                }
            }
        }
        
        System.out.println(cnt + " " + cnt2);
    }
    
    static void dfs(int x, int y, char color) {
        visited[x][y] = true;
        
        for (int[] delta : deltas) {
            int nx = x + delta[0];
            int ny = y + delta[1];
            
            if (nx >= 0 && nx < n && ny >= 0 && ny < n) {
                if (!visited[nx][ny] && graph[nx][ny] == color) {
                    dfs(nx, ny, color);
                }
            }
        }
    }
}
